### PR TITLE
Download silent

### PR DESCRIPTION
**Pull request by qyra**
Adding support for the silent=Bool keyword when calling download.
The default behaviour is silent, to match the behaviour in windows and since most scripts will be downloading relatively small files. When required, silent=false will show the full progress information in wget and curl.
This is primarly aimed at making download usable in async environments, without causing jumbled output.
The relevant documentation is also updated to note the new parameter, and that windows URLs must have the http:// suffix (wget / curl adds it automatically)

### CODE REVIEW

File: base/interactiveutil.jl [hunk 1]
```diff
@@ -373,9 +376,11 @@ end
     filename
 end
 
-function download(url::AbstractString)
+function download(url::AbstractString; silent::Bool=true)
     filename = tempname()
-    download(url, filename)
+    @unix_only download(url, filename, silent=silent)
+    @windows_only download(url, filename)
```
Comment (omus) [inline base/interactiveutil.jl]: Consider doing

``` julia
@unix ? download(url, filename, silent=silent) : download(url, filename)
```

File: base/interactiveutil.jl [hunk 2]
```diff
@@ -373,9 +376,11 @@ end
...
+    filename
```
Comment (omus) [inline base/interactiveutil.jl]: I'd remove this. Put a return on the `download` line if you want things to be clearer

File: base/docs/helpdb.jl [hunk 3]
```diff
@@ -864,7 +864,8 @@ takebuf_array
 doc"""
     download(url,[localfile])
 
-Download a file from the given url, optionally renaming it to the given local file name. Note that this function relies on the availability of external tools such as `curl`, `wget` or `fetch` to download the file and is provided for convenience. For production use or situations in which more options are needed, please use a package that provides the desired functionality instead.
+Download a file from the given url, optionally renaming it to the given local file name. Note that this function relies on the availability of external tools such as `curl`, `wget` or `fetch` to download the file and is provided for convenience. For production use or situations in which more options are needed, please use a package that provides the desired functionality instead. Platform specific details: In Linux / OS X with wget or curl, passing silent=false will display their progress messages. In windows, download requires a full URL, complete with the http:// prefix.
+
```
Comment (omus) [inline base/docs/helpdb.jl:868]: Break this up into multiple lines using a [92-character line break](http://julia.readthedocs.org/en/latest/manual/documentation/#documentation).

**Comment by omus** [comment]
@iamed2 the silent option only will change the behaviour on UNIX environments. Since `download` has `@windows_only` and `@unix_only` methods should the silent keyword be included in the windows function even though it does nothing?

I'm starting to lean towards including the keyword for the windows method just so the documentation can be consistent.

**Comment by iamed2** [comment]
What is the current behaviour on Windows?

**Comment by iamed2** [comment]
Even if the behaviour is different, the two OSs should have the same function signature.

**Comment by omus** [comment]
The `@windows_only download` method is always silent and never displays anything

**Comment by omus** [comment]
@akrasian I've changed my mind on the `@windows_only download` method signature. Include the silent keyword even though it doesn't modify the behaviour. That way you can change the documentation signature to be: `download(url, [filename]; silent=false)`

**Comment by iamed2** [comment]
I thought silent was going to be the default?

**Comment by omus** [comment]
Ah, I see that it currently is set to be `silent=true` as the default. If the default is set to `silent=false` then this would be an opt-in change rather than a change forced upon people. I lean towards defaulting silent to false. Any other opinions?

File: base/docs/helpdb.jl (hunk 3 above)
Comment (qyra) [inline base/docs/helpdb.jl:868]: I had just been looking at the nearby docs and seen they were on one line, so I was trying to be consistent. Changed regardless

**Comment by qyra** [comment]
Keeping the current behaviour of silent is good, but since Windows was already silent, I figure that having the same function call do the same thing in any OS is best. Additionally, if people are usually downloading small HTML files (which in my experience is more common) than silent downloads are less intrusive than needing to pass silent=true every time.

**Comment by qyra** [comment]
Should attempting to download verbose in windows raise an error? Or just operate normally?

**Comment by omus** [comment]
I only wanted silent downloads when performing `@async` downloads when the output becomes intermixed and useless. Normally I don't have an issue with downloads having output

**Comment by qyra** [comment]
Okay, I'll change the default to silent=false if that makes more sense. Easy enough to change later.

**Comment by iamed2** [comment]
I guess setting `silent=false` by default will ease acceptance of this PR due to not changing behaviour, and if people want a change they'll comment there.
The benefit of `silent=false` is visibility on when downloads fail. There may be people who want this.
Windows `silent=false` should continue; we want to facilitate cross-platform code without special cases while giving people as much control as possible.

**Comment by omus** [comment]
I'm also thinking that `silent` should be an optional parameter and not a keyword argument. I can be swayed either way though.

**Comment by qyra** [comment]
It seems to me that something like `download("URL", false)` is a little unreadable. (for now with only two args obviously it is fine, as long as you've seen the docs) Too many things that boolean could refer to, and especially if more options are added to download later that would get confusing. (It seems like the type of function likely to be extended)

**Comment by omus** [comment]
@akrasian solid point. I was mainly thinking about `download(url, filename, silent=false)`. Lets stick to keywords here

**Comment by omus** [comment]
Ok, summary of things to change:
- [ ] Add silent keyword to download method for windows
- [ ] Change silent keyword default to false

File: base/docs/helpdb.jl (hunk 3 above)
Comment (omus) [inline base/docs/helpdb.jl:868]: Understandable. These guides are rather new but all new changes are expected to abide by them

**Comment by iamed2** [comment]
- [ ] Make this PR against master instead of release-0.4

**Pull request closed without merging**